fiftyone/core/map/threading.py

```diff
@@ -120,121 +80,54 @@ def sample_map_worker(
-        progress: Optional[bool] = None,
-        save: Optional[bool] = None,
-    ) -> Iterator[Any]:
-        """Map samples in parallel"""
-
-        if num_workers is None:
-            # if fo.config.default_map_workers is not None:
-            #     num_workers = fo.config.default_map_workers
-            # else:
-            num_workers = fou.recommend_process_pool_workers()
-
-        if num_workers == 1:
-            yield from super().map_samples(
-                sample_collection,
-                map_fcn,
-                num_workers,
-                shard_method,
-                progress,
-                save,
-            )
-            return
-
-        sample_collections_and_counts = self.split_sample_collection(
-            sample_collection, num_workers, shard_method
-        )
-        sample_collection_count = len(sample_collections_and_counts)
-
+        progress: Union[bool, Literal["workers"]],
+        save: bool = False,
+    ) -> Iterator[Tuple[bson.ObjectId, R]]:
         q = queue.Queue()
         events: List[threading.Event] = []
 
+        batch_count = len(sample_batches)
         with concurrent.futures.ThreadPoolExecutor(
-            max_workers=sample_collection_count
+            max_workers=batch_count
         ) as executor:
             worker_progress = progress == "workers"
-            for idx, (sample_collection, sample_count) in enumerate(
-                sample_collections_and_counts
-            ):
-                events.append(threading.Event())
+
+            for idx, batch in enumerate(sample_batches):
+                batch_number = idx + 1
+
+                event = threading.Event()
+                events.append(event)
+
                 executor.submit(
-                    self.sample_map_worker,
-                    sample_collection,
-                    sample_count,
-                    idx + 1,
-                    sample_collection_count,
+                    self.__worker,
+                    batch.create_subset(self._sample_collection),
+                    batch.total,
+                    batch_number,
+                    batch_count,
                     q,
-                    events[-1],
-                    map_fcn=map_fcn,
-                    save=save,
-                    progress=worker_progress,
+                    event,
+                    map_fcn,
+                    save,
+                    worker_progress,
                 )
 
             def get_results(q, events):
                 while True:
                     try:
                         yield q.get(timeout=0.1)
                     except queue.Empty:
-                        events = [
-                            event for event in events if not event.is_set()
-                        ]
-
-                        if events:
-                            continue
-
-                        break
+                        events = [e for e in events if not e.is_set()]
```

Comment:
if there is an error in the individual `map_fcn`, the event might not be set. Should we add a max number of loop count here?

Reply:
I'm doing error handling separately right now. It'll be done in a follow up PR.